builtin/v9/migration/top.go

```diff
@@ -0,0 +1,341 @@
+
+func ActorHeadKey(addr address.Address, head cid.Cid) string {
+	headKey, err := head.StringOfBase(multibase.Base32)
+	if err != nil {
+		panic(err)
+	}
+
+	return addr.String() + "-head-" + headKey
+}
+
+// Migrates the filecoin state tree starting from the global state tree and upgrading all actor state.
+// The store must support concurrent writes (even if the configured worker count is 1).
+func MigrateStateTree(ctx context.Context, store cbor.IpldStore, newManifestCID cid.Cid, actorsRootIn cid.Cid, priorEpoch abi.ChainEpoch, cfg Config, log Logger, cache MigrationCache) (cid.Cid, error) {
+	if cfg.MaxWorkers <= 0 {
+		return cid.Undef, xerrors.Errorf("invalid migration config with %d workers", cfg.MaxWorkers)
+	}
+
+	adtStore := adt8.WrapStore(ctx, store)
+
+	// Load input and output state trees
+	actorsIn, err := LoadTree(adtStore, actorsRootIn)
+	if err != nil {
+		return cid.Undef, err
+	}
+	actorsOut, err := NewTree(adtStore)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	// load old manifest data
+	systemActor, ok, err := actorsIn.GetActor(builtin.SystemActorAddr)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to get system actor: %w", err)
+	}
+
+	if !ok {
+		return cid.Undef, xerrors.New("didn't find system actor")
+	}
+
+	var systemState system.State
+	if err := store.Get(ctx, systemActor.Head, &systemState); err != nil {
+		return cid.Undef, xerrors.Errorf("failed to get system actor state: %w", err)
+	}
+
+	var oldManifestData manifest.ManifestData
+	if err := store.Get(ctx, systemState.BuiltinActors, &oldManifestData); err != nil {
+		return cid.Undef, xerrors.Errorf("failed to get old manifest data: %w", err)
+	}
+
+	// load new manifest
+	var newManifest manifest.Manifest
+	if err := adtStore.Get(ctx, newManifestCID, &newManifest); err != nil {
+		return cid.Undef, xerrors.Errorf("error reading actor manifest: %w", err)
+	}
+
+	if err := newManifest.Load(ctx, adtStore); err != nil {
+		return cid.Undef, xerrors.Errorf("error loading actor manifest: %w", err)
+	}
+
+	// Maps prior version code CIDs to migration functions.
+	migrations := make(map[cid.Cid]actorMigration)
+
+	// Set of prior version code CIDs for actors to defer during iteration, for explicit migration afterwards.
+	var deferredCodeIDs = map[cid.Cid]struct{}{
+		// None
+	}
+
+	// simple code migrations
+	simpleMigrations := make(map[string]cid.Cid, len(oldManifestData.Entries))
+
+	for _, entry := range oldManifestData.Entries {
+		simpleMigrations[entry.Name] = entry.Code
+	}
+
+	for name, code7Cid := range simpleMigrations { //nolint:nomaprange
+		code8Cid, ok := newManifest.Get(name)
+		if !ok {
+			return cid.Undef, xerrors.Errorf("code cid for %s actor not found in manifest", name)
+		}
+
+		migrations[code7Cid] = codeMigrator{code8Cid}
+	}
+
+	// migrations that migrate both code and state
+	system8Cid, ok := newManifest.Get("system")
+	if !ok {
+		return cid.Undef, xerrors.Errorf("code cid for system actor not found in manifest")
+	}
```

Comment:
These and other things in this file should be updated for the new versions. If we expect to duplicate this code again in the future, it might be best to change names to `old` and `new`.

Reply:
yessir